taskchampion/lib/src/replica.rs

```diff
@@ -186,6 +235,156 @@ pub unsafe extern "C" fn tc_replica_new_on_disk(
+            },
+            ReplicaOp::Delete { .. } => TCReplicaOp {
+                operation_type: TCReplicaOpType::Delete,
+                inner: Box::new(replica_op),
+            },
+            ReplicaOp::Update { .. } => TCReplicaOp {
+                operation_type: TCReplicaOpType::Update,
+                inner: Box::new(replica_op),
+            },
+            ReplicaOp::UndoPoint => TCReplicaOp {
+                operation_type: TCReplicaOpType::UndoPoint,
+                inner: Box::new(replica_op),
+            },
+        }
+    }
+}
+
+#[ffizz_header::item]
+#[ffizz(order = 901)]
+/// ***** TCReplicaOpList *****
+///
+/// ```c
+/// struct TCReplicaOpList {
+///     struct TCReplicaOp *items;
+///     size_t len;
+///     size_t capacity;
+/// };
+///
+/// typedef struct TCReplicaOpList TCReplicaOpList;
+/// ```
+#[repr(C)]
+#[derive(Debug)]
+pub struct TCReplicaOpList {
+    items: *mut TCReplicaOp,
+    len: usize,
+    capacity: usize,
+}
+
+impl CList for TCReplicaOpList {
+    type Element = TCReplicaOp;
+
+    unsafe fn from_raw_parts(items: *mut Self::Element, len: usize, cap: usize) -> Self {
+        TCReplicaOpList {
+            len,
+            capacity: cap,
+            items,
+        }
+    }
+
+    fn slice(&mut self) -> &mut [Self::Element] {
+        // SAFETY:
+        //  - because we have &mut self, we have read/write access to items[0..len]
+        //  - all items are properly initialized Element's
+        //  - return value lifetime is equal to &mmut self's, so access is exclusive
+        //  - items and len came from Vec, so total size is < isize::MAX
+        unsafe { std::slice::from_raw_parts_mut(self.items, self.len) }
+    }
+
+    fn into_raw_parts(self) -> (*mut Self::Element, usize, usize) {
+        (self.items, self.len, self.capacity)
+    }
+}
+
+impl From<Vec<ReplicaOp>> for TCReplicaOpList {
+    fn from(replica_op_list: Vec<ReplicaOp>) -> TCReplicaOpList {
+        let tc_replica_op_vec: Vec<TCReplicaOp> = replica_op_list
+            .into_iter()
+            .map(|op| TCReplicaOp::from(op))
+            .collect();
+        let (ptr, len, capacity) = vec_into_raw_parts(tc_replica_op_vec);
+        TCReplicaOpList {
+            items: ptr as *mut TCReplicaOp,
+            len,
+            capacity,
+        }
+    }
+}
+
+impl From<TCReplicaOpList> for Vec<ReplicaOp> {
+    fn from(tc_replica_op_list: TCReplicaOpList) -> Vec<ReplicaOp> {
+        // SAFETY:
+        //  - items, len, and capacity are the unmodified values returned by vec_into_raw_parts
+        //  (promised by caller)
+        let tc_replica_op_vec = unsafe {
+            Vec::from_raw_parts(
+                tc_replica_op_list.items,
+                tc_replica_op_list.len,
+                tc_replica_op_list.capacity,
+            )
+        };
+
+        tc_replica_op_vec
+            .into_iter()
+            .map(|tc_op| *tc_op.inner)
+            .collect()
+    }
+}
```

Comment:
This is another `From` implementation that hides the unsafety. It can be removed with the suggested change to `tc_replica_commit_undo_ops`, below.